### PR TITLE
GH-8413: Execute the command if has active handler

### LINKED ISSUE #8413: [shell] Error when toggling the bottom panel with `Cltr/Cmd+j` when there isn't any views in the bottom panel
### Bug Description:

```
logger-protocol.ts:112 root ERROR Failed to execute command: Error: The command 'core.toggle.bottom.panel' cannot be executed. There are no active handlers available for the command. (args: [null])
    at CommandRegistry.<anonymous> (file:///Applications/Arduino%20Pro%20IDE.app/Contents/Resources/app/lib/bundle.js:57675:45)
    at step (file:///Applications/Arduino%20Pro%20IDE.app/Contents/Resources/app/lib/bundle.js:57444:23)
    at Object.next (file:///Applications/Arduino%20Pro%20IDE.app/Contents/Resources/app/lib/bundle.js:57425:53)
    at file:///Applications/Arduino%20Pro%20IDE.app/Contents/Resources/app/lib/bundle.js:57419:71
    at new Promise (<anonymous>)
    at ./node_modules/@theia/core/lib/common/command.js.__awaiter (file:///Applications/Arduino%20Pro%20IDE.app/Contents/Resources/app/lib/bundle.js:57415:12)
    at CommandRegistry../node_modules/@theia/core/lib/common/command.js.CommandRegistry.executeCommand (file:///Applications/Arduino%20Pro%20IDE.app/Contents/Resources/app/lib/bundle.js:57657:16)
    at KeybindingRegistry../node_modules/@theia/core/lib/browser/keybinding.js.KeybindingRegistry.executeKeyBinding (file:///Applications/Arduino%20Pro%20IDE.app/Contents/Resources/app/lib/bundle.js:35205:38)
    at KeybindingRegistry../node_modules/@theia/core/lib/browser/keybinding.js.KeybindingRegistry.run (file:///Applications/Arduino%20Pro%20IDE.app/Contents/Resources/app/lib/bundle.js:35297:22)
    at HTMLDocument.<anonymous> (file:///Applications/Arduino%20Pro%20IDE.app/Contents/Resources/app/lib/bundle.js:33430:35)
```

### Steps to Reproduce:

1. Make sure you do not have anything in the bottom panel.
2. Toggle it with the appropriate keybinding.
3.

### Additional Information

- Operating System:
- Theia Version:

### PR DESCRIPTION
**Pull request by kittaakos**
#### What it does

 - This PR makes sure the command is not executed via a keybinding if it has no active handlers.
 - Use `@theia/core/lib/common/keybinding` instead of `@theia/core/lib/browser/keybinding` to eliminate a few linter warnings.
 - Closes #8413.

#### How to test

 - Reset your workbench layout.
 - Toggle the bottom panel: `Ctrl/Cmd+J`.
 - Close the `Problems` view.
 - Toggle the bottom panel: `Ctrl/Cmd+J`
 - You must not see the error:
```
root ERROR Failed to execute command: Error: The command 'core.toggle.bottom.panel' cannot be executed. There are no active handlers available for the command. (args: [null])
    at CommandRegistry.<anonymous> (file:///Users/akos.kitta/git/theia/examples/electron/lib/bundle.js:134569:45)
    at step (file:///Users/akos.kitta/git/theia/examples/electron/lib/bundle.js:134338:23)
    at Object.next (file:///Users/akos.kitta/git/theia/examples/electron/lib/bundle.js:134319:53)
    at file:///Users/akos.kitta/git/theia/examples/electron/lib/bundle.js:134313:71
    at new Promise (<anonymous>)
    at ../../packages/core/lib/common/command.js.__awaiter (file:///Users/akos.kitta/git/theia/examples/electron/lib/bundle.js:134309:12)
    at CommandRegistry.../../packages/core/lib/common/command.js.CommandRegistry.executeCommand (file:///Users/akos.kitta/git/theia/examples/electron/lib/bundle.js:134551:16)
    at KeybindingRegistry.../../packages/core/lib/browser/keybinding.js.KeybindingRegistry.executeKeyBinding (file:///Users/akos.kitta/git/theia/examples/electron/lib/bundle.js:111317:38)
    at KeybindingRegistry.../../packages/core/lib/browser/keybinding.js.KeybindingRegistry.run (file:///Users/akos.kitta/git/theia/examples/electron/lib/bundle.js:111409:22)
    at HTMLDocument.<anonymous> (file:///Users/akos.kitta/git/theia/examples/electron/lib/bundle.js:109575:35)
```

 - Reset your workbench layout.
 - Open one or two editors.
 - Close the editors in the main area: `Ctrl/Cmd+K Ctrl/Cmd+W` in electron or `Alt+Shift+W` in browser.
 - Editors are closed.
 - Close the editors in the main area: `Ctrl/Cmd+K Ctrl/Cmd+W` in electron or `Alt+Shift+W` in browser.
 - You must not see the errors:
```
root ERROR Failed to execute command: Error: The command 'core.close.all.tabs' cannot be executed. There are no active handlers available for the command. (args: [null])
    at CommandRegistry.<anonymous> (file:///Users/akos.kitta/git/theia/examples/electron/lib/bundle.js:134569:45)
    at step (file:///Users/akos.kitta/git/theia/examples/electron/lib/bundle.js:134338:23)
    at Object.next (file:///Users/akos.kitta/git/theia/examples/electron/lib/bundle.js:134319:53)
    at file:///Users/akos.kitta/git/theia/examples/electron/lib/bundle.js:134313:71
    at new Promise (<anonymous>)
    at ../../packages/core/lib/common/command.js.__awaiter (file:///Users/akos.kitta/git/theia/examples/electron/lib/bundle.js:134309:12)
    at CommandRegistry.../../packages/core/lib/common/command.js.CommandRegistry.executeCommand (file:///Users/akos.kitta/git/theia/examples/electron/lib/bundle.js:134551:16)
    at KeybindingRegistry.../../packages/core/lib/browser/keybinding.js.KeybindingRegistry.executeKeyBinding (file:///Users/akos.kitta/git/theia/examples/electron/lib/bundle.js:111317:38)
    at KeybindingRegistry.../../packages/core/lib/browser/keybinding.js.KeybindingRegistry.run (file:///Users/akos.kitta/git/theia/examples/electron/lib/bundle.js:111409:22)
    at HTMLDocument.<anonymous> (file:///Users/akos.kitta/git/theia/examples/electron/lib/bundle.js:109575:35)
```

#### Review checklist

- [x] as an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)

#### Reminder for reviewers

- as a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

### CODE REVIEW

**Comment by akosyakov** [comment]
It breaks lazy activation of VS Code extensions on command. When a command is about to execute it fires `onWillExecuteCommand` event which gives a chance to participate other clients (like plugin system) and contribute handlers.

**Comment by kittaakos** [comment]
> breaks lazy activation of VS Code extensions on command

I did not know this. Any recommendations on how to fix the bug?

**Comment by akosyakov** [comment]
> I did not know this. Any recommendations on how to fix the bug?

Ignore it. Plugin system contributes the stub handler so it should be fine. Checked with python extension it is good.

File: packages/core/src/browser/keybinding.ts [hunk 1]
```diff
@@ -416,14 +416,16 @@ export class KeybindingRegistry {
...
-    protected executeKeyBinding(binding: Keybinding, event: KeyboardEvent): void {
+    protected executeKeyBinding(binding: common.Keybinding, event: KeyboardEvent): void {
         if (this.isPseudoCommand(binding.command)) {
             /* Don't do anything, let the event propagate.  */
         } else {
             const command = this.commandRegistry.getCommand(binding.command);
             if (command) {
-                this.commandRegistry.executeCommand(binding.command, binding.args)
-                    .catch(e => console.error('Failed to execute command:', e));
+                if (this.commandRegistry.isEnabled(binding.command, binding.args)) {
```
Comment (akosyakov) [inline packages/core/src/browser/keybinding.ts:425]: I only wonder whether it is a proper place. We have a method `isEnabled` in this class, which actually used to match keybindings. I am not sure whether it should be moved there? i.e. if keybinding is matching, but there is no active handler, should we skip keybinding and try next? With current proposal we pick keybinding and do nothing. There is also `isActive` method which is not used anymore. Maybe during refactoring we lost it and the actual fix is to call `isActive` from `isEnabled` if keybinding is matching.

Comment (kittaakos) [reply packages/core/src/browser/keybinding.ts:425]: Yes, this is fair reasoning.

> With current proposal we pick keybinding and do nothing

Well, with the `master` state, we pick the keybinding, and we equally do nothing as an error is thrown.

With the proposed change:
 - we do not fire the `fireWillExecuteCommand` if there are no active handlers. (same on `master`)
 - we `preventDefault` and `stopPropagation` no matter there were any active handlers or not. (same as on `master`)

The only behavioral change is that we do not throw an error.

> Maybe during refactoring we lost it

OK, I will look into the commit history.

Comment (akosyakov) [reply packages/core/src/browser/keybinding.ts:425]: ok, let's try

**Review by akosyakov — APPROVED** [review]
it seems to be minimal change

**Comment by vince-fugnitto** [comment]
@kittaakos @akosyakov do we merge for the release?

**Comment by kittaakos** [comment]
> @kittaakos @akosyakov do we merge for the release?

No.

**Pull request merged**